### PR TITLE
Issue #13321: Kill mutation for BlockCommentPosition 2

### PR DESCRIPTION
**Pull request by Kevin222004**
Issue #13321: Kill mutation for BlockCommentPosition 2

------

# Mutations 
https://github.com/checkstyle/checkstyle/blob/8804126fd3a4f58d5116712aaba939bcd2269b52/config/pitest-suppressions/pitest-utils-suppressions.xml#L21-L28

-----

# Explaination 

dependency tree on this method
```
BlockCommentPosition.isOnPlainClassMember
  +--isOnMethod
       +--ClassAndPropertiesSettersJavadocScraper
       +--BlockCommentPosition.isOnMember
            +--JavadocUtil.isCorrectJavadocPosition
                 +--InvalidJavadocPositionCheck
                 +--JavadocUtil.isJavadocComment
                      +--AstTreeStringPrinter
                      +--JavadocPropertiesGenerator
                      +--DesignForExtensionCheck
                      +--AbstractJavadocCheck
                           +-- all ast based javadoc checks
                      +--JavadocContentLocationCheck
                      +--MissingJavadocPackageCheck
  +--isOnField
     +-- isCorrectJavadocPosition ... same as above
  +--isOnConstructor
     +-- isCorrectJavadocPosition ... same as above
  +--isOnAnnotationField
     +-- isCorrectJavadocPosition ... same as above
```

so modules:
all ast based javadoc checks (MissingDeprecatedCheck, AtclauseOrderCheck, JavadocBlockTagLocationCheck, JavadocMissingLeadingAsteriskCheck, JavadocMissingWhitespaceAfterAsteriskCheck, JavadocParagraphCheck, JavadocTagContinuationIndentationCheck, NonEmptyAtclauseDescriptionCheck, RequireEmptyLineBeforeBlockTagGroupCheck, SingleLineJavadocCheck, SummaryJavadocCheck)
InvalidJavadocPositionCheck
DesignForExtensionCheck
MissingJavadocPackageCheck
JavadocContentLocationCheck

 we can reuse https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/checks-only-javadoc-error.xml 

-------

# Regression 

-----

Diff Regression config: https://gist.githubusercontent.com/romani/c02d1cfd4005bdd25a4a1c9d0e5ec09c/raw/235d6dd039b9b3ee0c01afce35be1771af3ecb55/pull-13424-regression-config.xml
Diff Regression projects: https://gist.githubusercontent.com/Kevin222004/21e3934e85f802e2fbd48af06d122364/raw/604256badd733d8568064f371d55657c04b00dfd/test-projects-2.properties
Report label: Final-Report-2

### CODE REVIEW

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
Github, generate report

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/BlockCommentPosition.java [hunk 1]
```diff
@@ -281,18 +281,16 @@ && getPrevSiblingSkipComments(blockComment.getParent()) == null
      * Checks that block comment is on specified class member without any modifiers.
      *
      * @param blockComment block comment start DetailAST
-     * @param memberType parent token type
      * @return true if block comment is on specified token without modifiers
      */
-    private static boolean isOnPlainClassMember(DetailAST blockComment, int memberType) {
+    private static boolean isOnPlainClassMember(DetailAST blockComment) {
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/BlockCommentPosition.java:286]: Javadoc of this method:

> /**
     * Checks that block comment is on specified class member without any modif

It is even strange why we need such javadoc in their most simplistic form.
We should support javadoc at any location.

**Comment by romani** [comment]
I am supporting removal of this condition. Let make PR ready for review

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/BlockCommentPosition.java (hunk 1 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/BlockCommentPosition.java:286]: https://checkstyle.org/checks/javadoc/invalidjavadocposition.html
Javadocs are not valid at every location.

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/BlockCommentPosition.java:286]: It also more importantly used to identify if we should process the comment as a javadoc and pass it to the javadoc checks.
https://github.com/checkstyle/checkstyle/blob/8f9c514d2f96bf9b9ef165931a6c509fe369c5c2/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheck.java#L310

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/BlockCommentPosition.java:286]: >Checks that block comment is on specified class member without any modifiers.

please change to `Checks that block comment is above all elements of class members`.


and this brings that we can try to kill it by javadoc comment on static block or any other 
that are not referenced at https://github.com/checkstyle/checkstyle/blob/8f9c514d2f96bf9b9ef165931a6c509fe369c5c2/src/main/java/com/puppycrawl/tools/checkstyle/utils/BlockCommentPosition.java#L141-L148
:
```
class A {
  /** some */
  { }
}
```

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/BlockCommentPosition.java:286]: this two conditons are enough for this method that test javadoc on class member:
```
                && !parent.getPreviousSibling().hasChildren()
                && parent.getParent().getParent().getType() == TokenTypes.OBJBLOCK;
```
if we make javadoc on non Method or Field .....:
```
    public static boolean isOnMember(DetailAST blockComment) {
        return isOnMethod(blockComment)
                || isOnField(blockComment)
                || isOnConstructor(blockComment)
                || isOnEnumConstant(blockComment)
                || isOnAnnotationField(blockComment)
                || isOnCompactConstructor(blockComment);
    }
```

can it be nested class?

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/BlockCommentPosition.java:286]: nested class is not plain member it is not affecting mutation.

**Comment by romani** [comment]
Github, generate report

**Comment by romani** [comment]
it is not a regresson:
```
/var/tmp$ cat config.xml 
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_2.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="DesignForExtensionCheck">
        </module>
    </module>
</module>

/var/tmp$ cat Test.java 
public class Test {
   void anotherMethod(String s);
   String field;
   /**
    * @param anObject
    * @param field
    */
   static void method(Test anObject, String field) {
       anObject.anotherMethod(field);
  }
}

/var/tmp$ java -jar checkstyle-10.12.4-all.jar -c config.xml Test.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing Test.java
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:309)
	at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:226)
	at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:415)
	at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:338)
	at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:195)
	at com.puppycrawl.tools.checkstyle.Main.main(Main.java:130)
Caused by: java.lang.NullPointerException
	at com.puppycrawl.tools.checkstyle.checks.design.DesignForExtensionCheck.hasEmptyImplementation(DesignForExtensionCheck.java:360)
	at com.puppycrawl.tools.checkstyle.checks.design.DesignForExtensionCheck.visitToken(DesignForExtensionCheck.java:256)
	at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:335)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:406)
	at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:273)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:158)
	at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:101)
	at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:337)
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:296)
	... 5 more
Checkstyle ends with 1 errors.

/var/tmp$ java -jar checkstyle-10.12.5-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
com.puppycrawl.tools.checkstyle.api.CheckstyleException: Exception was thrown while processing Test.java
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:311)
	at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:226)
	at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:415)
	at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:338)
	at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:195)
	at com.puppycrawl.tools.checkstyle.Main.main(Main.java:130)
Caused by: java.lang.NullPointerException
	at com.puppycrawl.tools.checkstyle.checks.design.DesignForExtensionCheck.hasEmptyImplementation(DesignForExtensionCheck.java:360)
	at com.puppycrawl.tools.checkstyle.checks.design.DesignForExtensionCheck.visitToken(DesignForExtensionCheck.java:256)
	at com.puppycrawl.tools.checkstyle.TreeWalker.notifyVisit(TreeWalker.java:335)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processIter(TreeWalker.java:408)
	at com.puppycrawl.tools.checkstyle.TreeWalker.walk(TreeWalker.java:273)
	at com.puppycrawl.tools.checkstyle.TreeWalker.processFiltered(TreeWalker.java:158)
	at com.puppycrawl.tools.checkstyle.api.AbstractFileSetCheck.process(AbstractFileSetCheck.java:101)
	at com.puppycrawl.tools.checkstyle.Checker.processFile(Checker.java:340)
	at com.puppycrawl.tools.checkstyle.Checker.processFiles(Checker.java:299)
	... 5 more
Checkstyle ends with 1 errors.

```

moved to https://github.com/checkstyle/checkstyle/issues/14022

**Review by romani — APPROVED** [review]
ok to merge

**Review by nrmancuso — APPROVED** [review]
It seems reasonable that we do not need to know the parent of the token that has the comment on it, as long as we know that it is in an OBJBLOCK.

**Pull request merged**